extension.js

```diff
@@ -174,6 +188,99 @@ class HuaweiWmiIndicator extends PanelMenu.Button { // TODO: move to system batt
 		this._bpm.label.set_text(this._BPM + `: ${low}%-${high}%`);
 	}
 
+	_update_top_off() {
+		this._get_battery(proxy => {
+			let is_discharging = (proxy.State === UPower.DeviceState.DISCHARGING)
+			let is_fully_charged = (proxy.State === UPower.DeviceState.FULLY_CHARGED)
+			if (is_fully_charged) this._stop_top_off();
+			else if (is_discharging) this._stop_top_off();
+		})
+	}
+
+	_start_top_off() {
+		let _file_sys = Gio.File.new_for_path(this._file_sys_str);
+
+		this._get_battery(proxy => {  // Connects watcher
+			this._battery_watching = proxy.connect('g-properties-changed', this._update_top_off.bind(this));
+			try {
+				_file_sys.replace_contents("0 100", null, false, 0, null);
+				this._topping_off = true;
+			} catch (e) {}
+		})
+	}
+
+	_stop_top_off() {
+		let _file_sys = Gio.File.new_for_path(this._file_sys_str);
+		let _file_def = Gio.File.new_for_path(this._file_def_str);
+
+		let def_low, def_high;
+		this._get_battery(proxy => {  // Disconnects watcher
+			proxy.disconnect(this._battery_watching);
+			try {  // Reinstates old BPM values
+				[def_low, def_high] = ByteArray.toString(_file_def.load_contents(null)[1]).split(' ').map(Number);
+				_file_sys.replace_contents(`${def_low} ${def_high}`, null, false, 0, null);
+				this._topping_off = false;
+			} catch (e) {}
+		});
+	}
+
+	_get_battery(callback) {
+		let menu = Main.panel.statusArea.aggregateMenu;
+		if (menu && menu._power) {
+			callback(menu._power._proxy, menu._power);
+		}
+	}
+
+	_set_top_off(state) {
+		let _file_sys = Gio.File.new_for_path(this._file_sys_str);
+		let _file_def = Gio.File.new_for_path(this._file_def_str);
+
+		let sys_low, sys_high, def_low, def_high;
+		let is_discharging;
+		this._get_battery(proxy => { is_discharging = (proxy.State === UPower.DeviceState.DISCHARGING) });
+
+		// Check if the button to enable battery top-off should be available and
+		// set toggle state depending on the actual values set in /sys and /etc
+		try {
+			[sys_low, sys_high] = ByteArray.toString(_file_sys.load_contents(null)[1]).split(' ').map(Number);
+			[def_low, def_high] = ByteArray.toString(_file_def.load_contents(null)[1]).split(' ').map(Number);
+
+			// If BPM == off -> Button = Unavailable and Off
+			if (def_low == 0 && def_high == 100) {
+				this._top_off.setToggleState(false);
+				this._top_off.setSensitive(false);
+			}
+			// If BPM is on and device is not discharging -> Button = Available
+			else if ((def_low != 0 || def_high != 100) && !is_discharging ) {
```

Comment:
Merge these to a single line too.

In general, don't place an `else` clause with a line break after the previous `}` it corresponds to, cause that's hard to follow.